### PR TITLE
SCMOD-4018: Update Marathon Autoscaler to use PATCH

### PR DESCRIPTION
**Pull request by rado-straka**
https://jira.autonomy.com/browse/SCMOD-4018
http://sou-jenkins2.hpeswlab.net/job/Autoscaler/view/Developer/job/Autoscaler~autoscaler~SCMOD-4018~CI/

### CODE REVIEW

File: autoscale-scaler-marathon/pom.xml [hunk 1]
```diff
@@ -60,6 +60,13 @@
             <artifactId>validation-api</artifactId>
             <scope>provided</scope>
         </dependency>
+
+        <!-- https://mvnrepository.com/artifact/org.apache.httpcomponents/httpclient -->
```
Comment (dermot-hardy) [inline autoscale-scaler-marathon/pom.xml]: Remove this comment - all open-source artifacts are in Maven Central (or The Central Repository as it calls itself these days).

File: autoscale-scaler-marathon/src/main/java/com/hpe/caf/autoscale/scaler/marathon/MarathonServiceScaler.java [hunk 2]
```diff
@@ -15,51 +15,52 @@
...
 import org.slf4j.LoggerFactory;
 
 import java.io.IOException;
 import java.net.InetSocketAddress;
 import java.net.Socket;
 import java.net.URL;
-import java.util.Collection;
-import java.util.LinkedList;
-import java.util.Objects;
+import java.util.*;
```
Comment (dermot-hardy) [inline autoscale-scaler-marathon/src/main/java/com/hpe/caf/autoscale/scaler/marathon/MarathonServiceScaler.java]: List out classes explicitly - don't use star imports.

File: autoscale-scaler-marathon/src/main/java/com/hpe/caf/autoscale/scaler/marathon/MarathonServiceScaler.java [hunk 3]
```diff
@@ -70,11 +71,10 @@ public void scaleUp(final String serviceReference, final int amount)
             int current = app.getTasksRunning() + app.getTasksStaged();
             int target = Math.min(maximumInstances, current + amount);
             if ( target > current ) {
-                app.setInstances(Math.min(maximumInstances, app.getTasksRunning() + app.getTasksStaged() + amount));
                 LOG.debug("Scaling service {} up by {} instances", serviceReference, amount);
-                marathon.updateApp(serviceReference, app, true);
+                appInstancePatcher.patchInstances(app.getId(), target);
             }
-        } catch (MarathonException e) {
+        } catch (Exception e) {
```
Comment (dermot-hardy) [inline autoscale-scaler-marathon/src/main/java/com/hpe/caf/autoscale/scaler/marathon/MarathonServiceScaler.java]: Don't catch the top-level `Exception`.  What type of exception are you actually trying to map?

File: autoscale-scaler-marathon/src/main/java/com/hpe/caf/autoscale/scaler/marathon/MarathonServiceScaler.java [hunk 4]
```diff
@@ -89,11 +89,10 @@ public void scaleDown(final String serviceReference, final int amount)
             App app = appGet.getApp();
             int current = app.getTasksRunning() + app.getTasksStaged();
             if ( current > 0 ) {
-                app.setInstances(Math.max(0, current - amount));
                 LOG.debug("Scaling service {} down by {} instances", serviceReference, amount);
-                marathon.updateApp(serviceReference, app, true);
+                appInstancePatcher.patchInstances(app.getId(), Math.max(0, current - amount));
             }
-        } catch (MarathonException e) {
+        } catch (Exception e) {
```
Comment (dermot-hardy) [inline autoscale-scaler-marathon/src/main/java/com/hpe/caf/autoscale/scaler/marathon/MarathonServiceScaler.java]: Same comment

File: autoscale-scaler-marathon/src/main/java/com/hpe/caf/autoscale/scaler/marathon/AppInstancePatcher.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+        final JsonObject details = new JsonObject();
+        details.addProperty("id", appId);
+        details.addProperty("instances", instances);
+
+        final JsonArray appArray = new JsonArray();
+        appArray.add(details);
+        
+        try(final CloseableHttpClient client = HttpClientBuilder.create().build()){
+            URIBuilder uriBuilder = new URIBuilder(new URL(marathonUri.toURL(), "/v2/apps").toURI());
```
Comment (dermot-hardy) [inline autoscale-scaler-marathon/src/main/java/com/hpe/caf/autoscale/scaler/marathon/AppInstancePatcher.java]: - `uriBuilder` should be `final`
- You jumped through hoops to pass `marathonUri` in as a `URI` only to have to get it back to a `URL`.

File: autoscale-scaler-marathon/src/main/java/com/hpe/caf/autoscale/scaler/marathon/AppInstancePatcher.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+            final HttpResponse response = client.execute(patch);
+            if(!force && response.getStatusLine().getStatusCode()==409){
+                patchInstances(appId, instances, true);
+                return;
+            }
+            if(response.getStatusLine().getStatusCode()!=200){
+                throw new ScalerException(response.getStatusLine().getReasonPhrase());
+            }
+        }
+        catch (Exception ex){
```
Comment (dermot-hardy) [inline autoscale-scaler-marathon/src/main/java/com/hpe/caf/autoscale/scaler/marathon/AppInstancePatcher.java]: This will catch unchecked exceptions.

**Review by dermot-hardy — CHANGES_REQUESTED** [review]

File: autoscale-scaler-marathon/src/main/java/com/hpe/caf/autoscale/scaler/marathon/AppInstancePatcher.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClientBuilder;
+
+import java.net.URI;
+import java.net.URL;
+import java.util.Arrays;
+import org.apache.http.client.utils.URIBuilder;
+import org.apache.http.message.BasicNameValuePair;
+
+public class AppInstancePatcher {
```
Comment (dermot-hardy) [inline autoscale-scaler-marathon/src/main/java/com/hpe/caf/autoscale/scaler/marathon/AppInstancePatcher.java]: Class should probably be `final`.

**Review by dermot-hardy — CHANGES_REQUESTED** [review]

**Comment by rado-straka** [comment]
Updated

**Review by dermot-hardy — APPROVED** [review]

**Pull request merged**